connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/RequestEvents.scala

```diff
@@ -0,0 +1,327 @@
+        case proto.Plan.OpTypeCase.COMMAND => request.getPlan.getCommand
+        case proto.Plan.OpTypeCase.ROOT => request.getPlan.getRoot
+        case _ =>
+          throw new UnsupportedOperationException(
+            s"${request.getPlan.getOpTypeCase} not supported.")
+      }
+
+    listenerBus.post(
+      SparkListenerConnectOperationStarted(
+        jobTag,
+        operationId,
+        clock.getTimeMillis(),
+        request.getSessionId,
+        request.getUserContext.getUserId,
+        request.getUserContext.getUserName,
+        Utils.redact(
+          sessionHolder.session.sessionState.conf.stringRedactionPattern,
+          ProtoUtils.abbreviate(plan, RequestEvents.MAX_STATEMENT_TEXT_SIZE).toString),
+        Some(request)))
+  }
+
+  /**
+   * Post @link org.apache.spark.sql.connect.service.SparkListenerConnectOperationAnalyzed.
+   *
+   * @param analyzedPlan
+   *   The analyzed plan generated by the Connect request plan. None when the request does not
+   *   generate a plan.
+   */
+  def postAnalyzed(analyzedPlan: Option[LogicalPlan] = None): Unit = {
+    val event =
+      SparkListenerConnectOperationAnalyzed(jobTag, operationId, clock.getTimeMillis())
+    event.analyzedPlan = analyzedPlan
+    listenerBus.post(event)
+  }
+
+  /**
+   * Post @link
+   * org.apache.spark.sql.connect.service.SparkListenerConnectOperationReadyForExecution.
+   */
+  def postReadyForExecution(): Unit = {
+    listenerBus.post(
+      SparkListenerConnectOperationReadyForExecution(jobTag, operationId, clock.getTimeMillis()))
+  }
+
+  /**
+   * Post @link org.apache.spark.sql.connect.service.SparkListenerConnectOperationCanceled.
+   */
+  def postCanceled(): Unit = {
+    listenerBus
+      .post(SparkListenerConnectOperationCanceled(jobTag, operationId, clock.getTimeMillis()))
+  }
+
+  /**
+   * Post @link org.apache.spark.sql.connect.service.SparkListenerConnectOperationFailed.
+   *
+   * @param errorMessage
+   *   The message of the error thrown during the request.
+   */
+  def postFailed(errorMessage: String): Unit = {
+    listenerBus.post(
+      SparkListenerConnectOperationFailed(
+        jobTag,
+        operationId,
+        clock.getTimeMillis(),
+        errorMessage))
+  }
+
+  /**
+   * Post @link org.apache.spark.sql.connect.service.SparkListenerConnectOperationFinished.
+   */
+  def postFinished(): Unit = {
+    listenerBus
+      .post(SparkListenerConnectOperationFinished(jobTag, operationId, clock.getTimeMillis()))
+  }
+
+  /**
+   * Post @link org.apache.spark.sql.connect.service.SparkListenerConnectOperationClosed.
+   */
+  def postClosed(): Unit = {
+    listenerBus
+      .post(SparkListenerConnectOperationClosed(jobTag, operationId, clock.getTimeMillis()))
+  }
+
+  def analyzed(tracker: QueryPlanningTracker, analyzedPlan: LogicalPlan): Unit = {
+    postAnalyzed(Some(analyzedPlan))
+  }
+
+  def readyForExecution(tracker: QueryPlanningTracker): Unit = postReadyForExecution
```

Comment:
hm, these are public just for the sake of being in the callback. could you create some small private util class for that instead of passing the whole RequestsEvents to the tracker? 
like
```
def getQueryPlanningTrackerCallback() = {
  new QueryPlanningTrackerCallback {
    def analyzed(tracker: QueryPlanningTracker, analyzedPlan: LogicalPlan): Unit = {
      postAnalyzed(Some(analyzedPlan))
    }

   def readyForExecution(tracker: QueryPlanningTracker): Unit = postReadyForExecution
}
```